### ISSUE TITLE
Authentication failure occurs on a multi-master environment.

### ISSUE THREAD

**Issue by kstreee** [saltstack/salt#42377]
### Description of Issue/Question

1. salt master generates its own key data and [writes to file](https://github.com/saltstack/salt/blob/2016.11/salt/daemons/masterapi.py#L208-L218) while starting. (ex, `.root_key`).
2. The written file will be used to build payloads for [CLI](https://github.com/saltstack/salt/blob/2016.11/salt/cli/salt.py#L75-L76), and others.
3. salt master compares master's key data (kept in memory) and payload's key data (read from the file) while [publishing payloads](https://github.com/saltstack/salt/blob/2016.11/salt/master.py#L2190).

The above execution scenario is working well if we don't share `cachedir` among multiple masters.
If we share the cache directory physically (i.e. mounting NAS, ...), several masters overwrite the key files while starting, and it causes 'authentication failure'. (because, the salt master's key data, which is kept in memory, and payload's key data, which is obtained from the file, could be different.)
Because of this issue, CLI is only available for the latest started salt master.

I think the base directory of key files ('.root_key' and other key files, like '{user}_key', ...) should be moved from 'cachedir' to the other directory. But, I'm not sure that moving base directory of key files is an appropriate solution for this issue, and I'm not sure which directory is appropriate for base directory of key files (`root`? `pkidir`? or making another option like `keydir`?). If you have any opinions about this, please leave comments.

### Setup

Set following master configuration for multiple masters.
```
...
root_dir: /
pki_dir: /home/user/salt/pki
sock_dir: /home/user/salt/sock
pidfile: /home/user/salt/pid/master
log_file: /home/user/salt/log/master
autosign_file: /home/user/salt/autoasign
autoreject_file: /home/user/salt/autoreject

# NAS mount point
cachedir: /mnt/salt/cache
...
```

### Steps to Reproduce Issue

Starting salt master one by one using the above master configuration, and execute simple salt CLI (sudo salt 'some-minion' cmd.run 'echo test', etc) on salt masters.
Only the latest started salt master can use salt CLI, but the other masters will fail to execute salt CLI by authentication failure.

Following is a specific 'authencation failure' scenario of this issue.
1. starting master 1
1-1. generating key data
1-2. writing the key data to '.root_key' on 'cachedir'
1-3. keeping the key data in master 1's memory.
2. starting master 2
2-1. generating key data
2-1. writing the key data to '.root_key' on 'cachedir' (**OVERRITE**)
2-2. keeping the key data in master 2's memory.
3. executing salt CLI on salt master 2 (working well)
4. executing salt CLI on salt master 1 (**authentication failure**)

### Versions Report

This bug(?) also exists in 'develop' branch of salt stack.

```
Salt Version:
           Salt: 2016.11.5-73-g3b493c6

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.6.0
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.12 (default, Dec 15 2016, 15:05:05)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.2
            ZMQ: 4.1.6

System Versions:
           dist: centos 6.7 Final
        machine: x86_64
        release: 2.6.32-696.3.2.el6.x86_64
         system: Linux
        version: CentOS 6.7 Final
```

**Comment by garethgreenaway** [comment]
Generally you don't want to share master keys between masters, each one should have it's own and the minions should have the ability to accept both.  I think this tutorial might be what you're looking to do:
https://docs.saltstack.com/en/latest/topics/tutorials/multimaster_pki.html

**Comment by kstreee** [comment]
@garethgreenaway Definitely I don't share master keys. In config, the 'pki_dir' directory already separated among multi-masters. The 'key' data in this issue is not referring pub-pri key of salt master or minion. it is different kinds of a key and I'm not sure the exact rules of the key, but [the key is generated on the 'cachedir'[ by deafult](https://github.com/saltstack/salt/blob/2016.11/salt/daemons/masterapi.py#L200-L206). For the 'key' data, I can't avoid sharing it because generation logic uses the 'cachedir' option.

**Comment by kstreee** [comment]
Is there any progress here? I'm not sure but, @DmitryKuzmenko may recognize about this issue because he did refactoring auth code recently which are related to this issue (f009097a9c2493014d0caea9dd091aa759bd0d7e). [This line of code](https://github.com/saltstack/salt/commit/f009097a9c2493014d0caea9dd091aa759bd0d7e#diff-42d832c88242dd1539ef50c1830ff564R298) is comparing 'key data from file' and 'key data from memory'.

**Comment by DmitryKuzmenko** [comment]
@kstreee I've just refactored the code keeping the original functionality. I think it's better to address your question to @cachedout, @terminalmage or @thatch45

**Comment by kstreee** [comment]
The pr #42532 is related to this issue (not pr for fixing this issue). It may help to understand this issue, and the role of 'key' files.

**Comment by kstreee** [comment]
This issue is resolved by the pr #43474
